node/core/dispute-participation/src/lib.rs

```diff
@@ -0,0 +1,357 @@
+		CandidateValidationMessage, DisputeCoordinatorMessage, DisputeParticipationMessage,
+		RuntimeApiMessage, RuntimeApiRequest,
+	},
+	ActiveLeavesUpdate, FromOverseer, OverseerSignal, SpawnedSubsystem, Subsystem,
+	SubsystemContext, SubsystemError,
+};
+use polkadot_primitives::v1::{BlockNumber, CandidateHash, CandidateReceipt, Hash, SessionIndex};
+
+const LOG_TARGET: &str = "parachain::dispute-participation";
+
+struct State {
+	recent_block: Option<(BlockNumber, Hash)>,
+}
+
+/// An implementation of the dispute participation subsystem.
+pub struct DisputeParticipationSubsystem;
+
+impl DisputeParticipationSubsystem {
+	/// Create a new instance of the subsystem.
+	pub fn new() -> Self {
+		DisputeParticipationSubsystem
+	}
+}
+
+impl<Context> Subsystem<Context> for DisputeParticipationSubsystem
+where
+	Context: SubsystemContext<Message = DisputeParticipationMessage>,
+{
+	fn start(self, ctx: Context) -> SpawnedSubsystem {
+		let future = run(ctx).map(|_| Ok(())).boxed();
+
+		SpawnedSubsystem {
+			name: "dispute-participation-subsystem",
+			future,
+		}
+	}
+}
+
+#[derive(Debug, thiserror::Error)]
+#[allow(missing_docs)]
+pub enum Error {
+	#[error(transparent)]
+	RuntimeApi(#[from] RuntimeApiError),
+
+	#[error(transparent)]
+	Subsystem(#[from] SubsystemError),
+
+	#[error(transparent)]
+	Oneshot(#[from] oneshot::Canceled),
+
+	#[error(transparent)]
+	Participation(#[from] ParticipationError),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ParticipationError {
+	#[error("Missing recent block state to participate in dispute.")]
+	MissingRecentBlockState,
+	#[error("Failed to recover available data.")]
+	MissingAvailableData,
+}
+
+impl Error {
+	fn trace(&self) {
+		match self {
+			// don't spam the log with spurious errors
+			Self::RuntimeApi(_) | Self::Oneshot(_) => {
+				tracing::debug!(target: LOG_TARGET, err = ?self)
+			}
+			// it's worth reporting otherwise
+			_ => tracing::warn!(target: LOG_TARGET, err = ?self),
+		}
+	}
+}
+
+async fn run<Context>(mut ctx: Context)
+where
+	Context: SubsystemContext<Message = DisputeParticipationMessage>,
+{
+	let mut state = State { recent_block: None };
+
+	loop {
+		match ctx.recv().await {
+			Err(_) => return,
+			Ok(FromOverseer::Signal(OverseerSignal::Conclude)) => {
+				tracing::info!(target: LOG_TARGET, "Received `Conclude` signal, exiting");
+				return;
+			}
+			Ok(FromOverseer::Signal(OverseerSignal::BlockFinalized(_, _))) => {}
+			Ok(FromOverseer::Signal(OverseerSignal::ActiveLeaves(update))) => {
+				update_state(&mut state, update);
+			}
+			Ok(FromOverseer::Communication { msg }) => {
+				if let Err(err) = handle_incoming(&mut ctx, &mut state, msg).await {
+					err.trace();
+					if let Error::Subsystem(SubsystemError::Context(_)) = err {
+						return;
+					}
+				}
+			}
+		}
+	}
+}
+
+fn update_state(state: &mut State, update: ActiveLeavesUpdate) {
+	for active in update.activated {
+		if Some(active.number) > state.recent_block.map(|s| s.0) {
```

Comment:
this is a nice trick, but maybe that is more readable
```suggestion
		if state.recent_block.map_or(true, |s| active.number > s.0) {
```